sdk/go/common/resource/plugin/rpc.go

```diff
@@ -168,6 +168,9 @@ func MarshalPropertyValue(v resource.PropertyValue, opts MarshalOptions) (*struc
 			val := string(ref.URN)
 			if ref.HasID {
 				val = string(ref.ID)
+				if val == "" {
```

Comment:
The semantics here are very subtle. Can we add comments somewhere appropriate that describe what these cases mean?

In particular - what ones hasid=true and id="" mean?

Reply:
Yeah, definitely.

The particular case that you've called out means "this reference has an ID, but that ID is unknown". In retrospect, it was probably a mistake to represent the ID as a simple string in the Go SDK rather than a `PropertyValue`, but I did want to be consistent with the semantics of resource IDs themselves (where an empty ID means that the ID is unknown). We need `HasID` to distinguish between refs that have an empty ID b/c the referenced resource is not a custom resource and refs that have an empty ID b/c the ID is unknown.

Reply:
Refactored for an approach that is hopefully clearer by construction.